relay-sampling/src/evaluation.rs

```diff
@@ -7,226 +7,205 @@ use chrono::{DateTime, Utc};
-    let rules = merge_rules_from_configs(sampling_config, root_sampling_config);
-    let mut match_result = SamplingMatch::match_against_rules(rules, event, dsc, now)?;
-
-    // If we have a match, we will try to derive the sample rate based on the sampling mode.
-    //
-    // Keep in mind that the sample rate received here has already been derived by the matching
-    // logic, based on multiple matches and decaying functions.
-    //
-    // The determination of the sampling mode occurs with the following priority:
-    // 1. Non-root project sampling mode
-    // 2. Root project sampling mode
-    let Some(primary_config) = sampling_config.or(root_sampling_config) else {
-        relay_log::error!("cannot sample without at least one sampling config");
-        return None;
-    };
-    let sample_rate = match primary_config.mode {
-        SamplingMode::Received => match_result.sample_rate,
-        SamplingMode::Total => match dsc {
-            Some(dsc) => dsc.adjusted_sample_rate(match_result.sample_rate),
-            None => match_result.sample_rate,
-        },
-        SamplingMode::Unsupported => {
-            if processing_enabled {
-                relay_log::error!("found unsupported sampling mode even as processing Relay");
+    rule_ids: Vec<RuleId>,
+    factor: f64,
+    client_sample_rate: Option<f64>,
+}
+
+impl SamplingEvaluator {
+    /// Constructor for [`SamplingEvaluator`].
+    pub fn new(now: DateTime<Utc>) -> Self {
+        Self {
+            now,
+            rule_ids: vec![],
+            factor: 1.0,
+            client_sample_rate: None,
+        }
+    }
+
+    /// Sets a new client sample rate value.
+    pub fn adjust_client_sample_rate(mut self, client_sample_rate: Option<f64>) -> Self {
+        self.client_sample_rate = client_sample_rate;
+        self
+    }
+
+    /// Attemps to find a match for sampling rules.
+    pub fn match_rules<'a, I, G>(mut self, seed: Uuid, instance: &G, rules: I) -> RuleMatchingState
+    where
+        G: Getter,
+        I: Iterator<Item = &'a SamplingRule>,
+    {
+        for rule in rules {
+            if !rule.condition.matches(instance) {
+                continue;
+            };
+
+            let Some(sampling_value) = rule.sample_rate(self.now) else {
+                continue;
+            };
+
+            self.rule_ids.push(rule.id);
+
+            match sampling_value {
+                SamplingValue::Factor { value } => self.factor *= value,
+                SamplingValue::SampleRate { value } => {
+                    let sample_rate = (value * self.factor).clamp(0.0, 1.0);
+
+                    return RuleMatchingState::SamplingMatch(SamplingMatch::new(
+                        self.adjusted_sample_rate(sample_rate),
+                        seed,
+                        self.rule_ids,
+                    ));
+                }
             }
+        }
+        RuleMatchingState::Evaluator(self)
```

Comment:
if i understand it, this kinda pattern doesnt make new allocations, but please let me know if thats not the case